highfidelity/hifi
# Use JSON for application Settings

Given the way the ini "file format" (There is no actually specified format) works and the way Qt decided to interpret it means that integer value and boolean don't retain their type past the life of the process. Any subsequent read will return a String.

To address that I wrote a QSetting format that uses JSON which is strongly typed and encodes non trivial types into a string (The same way Qt does in ini files).
- Switched Settings to use a JSON file
- Added a converter to get settings from ini files if no json file is found
- Cleaned up uses of QSettings throughout the codebase.
- Removed old backward compatible code that is no longer needed.
- Set Application Name, Organisation, Version earlier in it's lifetime (This is used by QSettings, QDebug and other services to determine where to put their files).
- Simplified Setting init process (only one call to `init` needed now, no more `preInit`)
- Removed ApplicationInfo.ini and moved values to BuildInfo.h
- Fixed an old bug that made it so the interface.ini file was shared by all installs (Now stable, dev and PRs each have their own .ini file)
# Fixes:

https://highfidelity.fogbugz.com/f/cases/1033/Boolean-Settings-values-are-sometimes-read-as-type-Boolean-and-other-times-as-type-string
https://highfidelity.fogbugz.com/f/cases/1514/Settings-getValue-of-boolean-gives-string-in-NEXT-session
# Test plan:
- Run Interface and Sandbox. (This should look like a normal first run)
- Close both.
- Go to `C:\Users\<username>\AppData\Roaming\High Fidelity`
- Copy `interface.ini` and `domain-server.ini`
- Paste them in `C:\Users\<username>\AppData\Roaming\High Fidelity - PR8789`
- Run Interface and Sandbox. (This should look like a normal first run)
- In `C:\Users\<username>\AppData\Roaming\High Fidelity - PR8789` `interface.json` and `domain-server.json` should appear.
- Compare them to their `.ini` counter part to make sure the settings were ported correctly.
- Play around with changing settings and restarting interface, make sure the settings persist correctly.
- Compare framerates in the main domains between the current Master and this PR.
# Test Bug Fix:
### On current Master:
- Starting with a clean .ini file
- Open JS Console
- Paste this and press enter:

``` js
function write(name, value) {
    Settings.setValue(name, value);
    print("Storing (" +  typeof value + ") " + name + " => " + value);
}

function read(name) {
    var value = Settings.getValue(name);
    print("Reading: (" +  typeof value + ") " + name + " => " + value);
}

write("IntInt", 10);
write("IntString", "10");
write("BoolBool", true);
write("BoolString", "true");
write("String", "This is a string");

read("IntInt");
read("IntString");
read("BoolBool");
read("BoolString");
read("String");
```
- You should observe the following output:

```
Storing (number) IntInt => 10
Storing (string) IntString => 10
Storing (boolean) BoolBool => true
Storing (string) BoolString => true
Storing (string) String => This is a string
Reading: (number) IntInt => 10
Reading: (string) IntString => 10
Reading: (boolean) BoolBool => true
Reading: (string) BoolString => true
Reading: (string) String => This is a string
```
- Notice the type read match the type written
- Restart interface
- Run the above JS code again
- You should observe the following output:

```
Storing (number) IntInt => 10
Storing (string) IntString => 10
Storing (boolean) BoolBool => true
Storing (string) BoolString => true
Storing (string) String => This is a string
Reading: (string) IntInt => 10
Reading: (string) IntString => 10
Reading: (string) BoolBool => true
Reading: (string) BoolString => true
Reading: (string) String => This is a string
```
- Notice all the types have been collapsed to string
### On the PR build:
- Execute the above step and notice the types are conserved.
## Why the heck is that happening?

It's a combination of how QSettings locally keeps the types in RAM and how Qt implements the ini format.
Here is what the ini format looks like:

``` ini
IntInt=10
IntString=10
BoolBool=true
BoolString=true
String=This is a string
```

Notice ints, bools and string are all stored the exact same way.
And on read of the ini file, Qt interprets all\* types as string.
Another thing is that once a key is read/set into QSettings, it keeps it's current type until the end of the process.
Meaning that if I try to set a key to a boolean after it read a string, it'll stay a string.
## Why have we not noticed that before?

Most of our settings are used in the C++ code. The C++ statically associates a type to a key. So the value get converted all the time to the expected type. This doesn't happen with JS.
## Why JSON

Several reasons, first, JSON is strongly typed, it can differentiate between numbers, bools and strings.
Second, it was much easier to write a JSON read/write format than trying to fix and hack the INI format.

Here is how JSON stores the data:

``` json

{
    "IntInt": 10,
    "IntString": "10",
    "BoolBool": true,
    "BoolString": "true",
    "String": "This is a string",
}
```
## Misc

On top of that, this makes our settings slightly more readable.
1. Everything is sorted alphabetically
2. We can have spaces!
3. Support more native types:
interface.ini:

``` ini
; QRect
WindowGeometry=@Rect(39 201 1391 735)
; double
[Overlay.AvatarInputs]
y=11.343201518058777
; float
[Razer%20Hydra]
avatarPosition\x=@Variant(\0\0\0\x87\xbe\x80\0\0)
; QUrl
[AddressManager]
address="@Variant(\0\0\0\x11\0\0\0=hifi://welcome/-15.936,0.446346,-6.63038/0,0.85065,0,0.525733)"
```

interface.json:

``` json
{
    // QRect
    "WindowGeometry": "@Rect(17 230 1309 749)",
    // double
    "Overlay.AvatarInputs/y": 1.3715223073959351,
    // float
    "Razer Hydra/avatarPosition/x": -0.25,
    // QUrl
    "AddressManager/address": "hifi://welcome/-15.5723,0.446328,-6.88071/0,0.838931,0,0.544238",
}
```

* Comments in the above examples are not in the actual files.
